feat(providers): add BoxyHQ SAML Jackson provider (#3782)
nextauthjs/next-auth: packages/docs/docs/providers/saml-jackson.md

---
id: saml-jackson
title: BoxyHQ SAML Jackson
---

## Documentation

BoxyHQ SAML Jackson is an open source service that handles the SAML login flow as an OAuth 2.0 flow, abstracting away all the complexities of the SAML protocol.

> You can deploy SAML Jackson as a separate service or embed it into your app using our NPM library. [Please check our documentation for more details](https://boxyhq.com/docs/jackson/deploy)

Reviewer: ```suggestion
You can deploy SAML Jackson as a separate service or embed it into your app using our NPM library. [Check out the documentation for more details](https://boxyhq.com/docs/jackson/deploy)
```

## Configuration

SAML login requires a configuration for every tenant of yours. One common method is to use the domain for an email address to figure out which tenant they belong to. You can also use a unique tenant ID (string) from your backend for this, typically some kind of account or organization ID.

> Check our [documentation](https://boxyhq.com/docs/jackson/getting-started#2-saml-config-api) for more details.

Reviewer: ```suggestion
Check out the [documentation](https://boxyhq.com/docs/jackson/getting-started#2-saml-config-api) for more details.
```

## Options

The **SAML Jackson Provider** comes with a set of default options:

> - [BoxyHQ Provider options](https://github.com/nextauthjs/next-auth/blob/main/src/providers/boxyhq.ts)

Reviewer: ```suggestion
- [BoxyHQ Provider options](https://github.com/nextauthjs/next-auth/tree/main/packages/next-auth/src/providers/boxyhq.ts)
```

Let's keep the filename/id the same everywhere. You can decide on which one to use. `boxyhq` or `saml-jackson` or something else?

You can override any of the options to suit your own use case.

## Example

```ts
import SAMLJacksonProvider from "next-auth/providers/saml-jackson"
...
providers: [
  SAMLJacksonProvider({

>     issuer: "http://localhost:5000",
>     clientId: "dummy", // The dummy here is necessary since we'll pass tenant and product custom attributes in the client code
>     clientSecret: "dummy", // The dummy here is necessary since we'll pass tenant and product custom attributes in the client code

Reviewer: ```suggestion
    issuer: "http://localhost:5000",
    clientId: "dummy", // The dummy here is necessary since we'll pass tenant and product custom attributes in the client code
    clientSecret: "dummy", // The dummy here is necessary since we'll pass tenant and product custom attributes in the client code
```

This is an unusual flow, so I would like to have more context here.

If the `clientId` and `clientSecret` values aren't important, we should just omit them, or set them by default in the provider code.

>   SAMLJacksonProvider({
>     issuer: "http://localhost:5000",
>     clientId: "dummy", // The dummy here is necessary since we'll pass tenant and product custom attributes in the client code
>     clientSecret: "dummy", // The dummy here is necessary since we'll pass tenant and product custom attributes in the client code
>   })

Reviewer: ```suggestion
  SAMLJacksonProvider({ issuer: process.env.SAML_JACKSON_ISSUER })
```

This is an unusual flow, so I would like to have more context here. Why are client id and client secret "dummy" values?

If the `clientId` and `clientSecret` values aren't important somehow (very curious), we should just omit them, or set them by default in the provider code.

From the localhost issuer, I will also assume that you don't have a hosted service/it's something the user will always need to set up themselves? Do you have something similar to https://demo.identityserver.io/ that the user could play with? It would be a nice mention on this docs page.

Is this referring to the same as here: https://github.com/nextauthjs/next-auth/pull/3782/files#diff-e4ec1563a89a51e2b61e12a89fc66bb226e55e3f27173919bdb9b9f75a713d85R10

Author: `clientId` and `clientSecret` aren't as critical here (unlike OAuth 2.0) because SAML is tenanted (unique config per customer, that is also configured by the customer using their SAML Identity Provider). When `tenant` and `product` are used then `clientId` and `clientSecret` become merely artifacts for the OAuth 2.0 flow. More details explained [here](https://boxyhq.com/docs/jackson/security#client-secret).

We just released a mock SAML service that can be used for testing - https://mocksaml.com, I'll add it to the docs. Please also try our demo here to get an understanding of the flow - https://saml-demo.boxyhq.com/

Author: I had originally internalised the `dummy` logic here - https://github.com/nextauthjs/next-auth/pull/3782/commits/8955ffd887a63fc40b670ed4251488e2c608e177#diff-76f06144da0418fa936f7f4af37fcf4b1c21448f94716360feef15e4c2be4013L46

Author: @balazsorban44 I have setup a demo that requires no configuration from your side. Please run the app from this branch - https://github.com/nextauthjs/next-auth/pull/3782/commits/afbdc5b4aee33ccc4ef1f5292ca44217a5fe94c3

}
...
```

On the client side you'll need to pass additional parameters `tenant` and `product` to the `signIn` function. This will allow SAML Jackson to figure out the right SAML configuration and take your user to the right SAML Identity Provider to sign them in.

```tsx

> import { getCsrfToken, signIn } from "next-auth/react";

Reviewer: ```suggestion
import { signIn } from "next-auth/react";
```

...

  // Map your users's email to a tenant and product
  const tenant = email.split("@")[1];

>   const product= 'my_awesome_product';

Reviewer: ```suggestion
  const product = 'my_awesome_product';
```

Could you explain where this value comes from? Is it static for the app's lifetime? Is it coming from a dropdown/select?

Author: `tenant` and `product` are decided by the user, typically unique values that identify their customer. SAML (unlike OAuth 2.0) is tenanted per customer (and product).

...
  <Button
    onClick={async (event) => {
      event.preventDefault();

>       signIn("saml", {}, { tenant: tenant, product: product });

Reviewer: ```suggestion
      signIn("saml-jackson", {}, { tenant, product });
```

The first argument must match the provider implementation's `id`. You can choose another name, but let's keep it consistent.

Author: Makes sense, I'll make this consistent.
